runtime/astar/src/precompiles.rs

```diff
@@ -34,39 +37,47 @@ impl<R> AstarNetworkPrecompiles<R> {
-        + Dispatchable<PostInfo = PostDispatchInfo>
-        + GetDispatchInfo
-        + Decode,
+    R::Call: Dispatchable<PostInfo = PostDispatchInfo> + GetDispatchInfo + Decode,
 {
     fn execute(
+        &self,
         address: H160,
         input: &[u8],
         target_gas: Option<u64>,
         context: &Context,
-    ) -> Option<Result<PrecompileOutput, ExitError>> {
+        is_static: bool,
+    ) -> Option<PrecompileResult> {
         match address {
             // Ethereum precompiles :
-            a if a == hash(1) => Some(ECRecover::execute(input, target_gas, context)),
-            a if a == hash(2) => Some(Sha256::execute(input, target_gas, context)),
-            a if a == hash(3) => Some(Ripemd160::execute(input, target_gas, context)),
-            a if a == hash(4) => Some(Identity::execute(input, target_gas, context)),
-            a if a == hash(5) => Some(Modexp::execute(input, target_gas, context)),
-            a if a == hash(6) => Some(Bn128Add::execute(input, target_gas, context)),
-            a if a == hash(7) => Some(Bn128Mul::execute(input, target_gas, context)),
-            a if a == hash(8) => Some(Bn128Pairing::execute(input, target_gas, context)),
+            a if a == hash(1) => Some(ECRecover::execute(input, target_gas, context, is_static)),
+            a if a == hash(2) => Some(Sha256::execute(input, target_gas, context, is_static)),
+            a if a == hash(3) => Some(Ripemd160::execute(input, target_gas, context, is_static)),
+            a if a == hash(4) => Some(Identity::execute(input, target_gas, context, is_static)),
+            a if a == hash(5) => Some(Modexp::execute(input, target_gas, context, is_static)),
+            a if a == hash(6) => Some(Bn128Add::execute(input, target_gas, context, is_static)),
+            a if a == hash(7) => Some(Bn128Mul::execute(input, target_gas, context, is_static)),
+            a if a == hash(8) => Some(Bn128Pairing::execute(input, target_gas, context, is_static)),
             // nor Ethereum precompiles :
-            a if a == hash(1024) => Some(Sha3FIPS256::execute(input, target_gas, context)),
-            a if a == hash(1025) => Some(Dispatch::<R>::execute(input, target_gas, context)),
-            a if a == hash(1026) => Some(ECRecoverPublicKey::execute(input, target_gas, context)),
+            a if a == hash(1024) => {
+                Some(Sha3FIPS256::execute(input, target_gas, context, is_static))
+            }
+            a if a == hash(1025) => Some(Dispatch::<R>::execute(
+                input, target_gas, context, is_static,
+            )),
+            a if a == hash(1026) => Some(ECRecoverPublicKey::execute(
+                input, target_gas, context, is_static,
+            )),
             // Astar precompiles (starts from 0x5000):
             // Default
             _ => None,
         }
     }
+
+    fn is_precompile(&self, address: H160) -> bool {
+        Self::used_addresses().find(|x| x == &address).is_some()
+    }
```

Comment:
Just an observation/question - I'm curious how come that we re-initialize the Vector in `used_addresses()` function and hash each value repeatedly for each call?
Seems fairly inefficient.

I assume it's because it's a function on associated type?

Reply:
You right, hope optimizer can deal with it.